### PR TITLE
add $DPATH environment variable for file locations

### LINKED ISSUE #229: Support custom paths for the default installation folder
Currently all packages are installed to `~/.dub/packages/*` by default, or `%AppData%\dub\packages\*` on Windows. A setting in the `~/dub/settings.json` fie should be made available to change that default.

### PR DESCRIPTION
**Pull request by WebFreak001**
fix #229

I chose not to add a setting for this to make it easier for other apps
in the ecosystem to understand how to locate and change the dub folder,
to parse packages, manipulate them, maintain them, etc.

Having a setting would require other users to potentially locate the
config files (which has a bunch of exceptions), parse a JSON and find
an optional value. Limiting it to an environment variable means that
also tools like scripts can easily find the location of this.

### CODE REVIEW

**Comment by Geod24** [comment]
But an environment variable is prone to other kind of issues. For one it's not persistent unless you edit your shell config, which looses locality. Also names are prone to collision. IMO if one wants to get this value in script, we should have a `dub config` command.

**Comment by WebFreak001** [comment]
imo if you set your environment variable only locally then you probably only want DUB to store its data where you specified in that session. (maybe set to `/tmp/dub` to quickly try out some package without downloading it, or different stores for every different user)

**Comment by WebFreak001** [comment]
if we don't plan to add other tools, it might also make sense to define a `DUB_HOME` environment variable instead, which defaults to `~/.dub` / `%APPDATA%/dub`, so the folder can be overwritten in whole.

File: source/dub/dub.d
Comment (WebFreak001) [inline source/dub/dub.d:1960]: should the setting for this be called `dubHome` or something else?

**Comment by WebFreak001** [comment]
after discussing with @Geod24 I changed this to also include a setting instead of only an environment variable. This is so that system-wide changes can be made, e.g. by linux package maintainers, but also kept an environment variable so that it's possible to create a quick temporary clean local dub package storage or in general if it might be needed for some specific user's projects on the system.

To support changing the path of the packages (and also the config), the config in /etc/dub/settings.json (or /usr/...) will be read first and if the `dubHome` property is set in there, loading of further configs will be redirected to there, as well as the packages being stored there.

**Comment by dk949** [comment]
Hi 👋

I've been working on something similar (I guess now I'll just wait for this to merge), and just wanted to share some thoughts from my implementation.

I also considered something like `DUB_HOME` or `DUB_PATH`, but came to the conclusion that it's far too generic, which means users may have it set to control some other part of their system (`DPATH` would suffer the most from this). 

I am not sure how common this is on Windows or Mac, but a lot of Linux programs use environment variables to control the location of their config/cache/etc files (see [this page](https://wiki.archlinux.org/title/XDG_Base_Directory#Partial) for some examples).

In the end I settled for `DUB_SYSTEM_SETTINGS_PATH`, `DUB_USER_SETTINGS_PATH` and `DUB_LOCAL_REPOSITORY_PATH`. As I understand you intend other D tooling to use this directory, so maybe something like `DLANG_HOME` (in line with `JDK_HOME`)?

As for "...it's not persistent unless you edit your shell config", well yes, but users are already used to doing this. zsh even has a special config file pretty much for this purpose (`.zshenv`).

[My impl](https://github.com/dk949/dub/tree/user_specified_directories) for reference.

**Comment by WebFreak001** [comment]
I made it now so that environment variables can only change both folders at once, but you can use the settings for finer control of user settings and repository path, or use the general "dubHome" setting to set both.

Additionally for package maintainers and sysadmins it's probably interesting to have environment variables in the settings, for example to have a different user store for each user, so I added that as well.

See the changelog entry for details https://github.com/dlang/dub/blob/5addac0e22036cf6b328bda72ac87c47d55d0331/changelog/dpath.dd

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:329]: The circular dependency is IMO weird. I need a bit of time to think about it.

**Review by Geod24 — COMMENTED** [review]
I just realized we read `/etc/` before `/var/lib`...

The idea of a "dub home" is a bit weird: We don't have a unified "root location". We have the badly-named `Repository` (see the `private struct` in `packagemanager.d`) and we tackled on config files on top of that.

Also what's the relationship with `customCachePaths` ?

**Comment by Geod24** [comment]
Any chance we could get https://github.com/dlang/dub/pull/2343 before, since it overhaul this part of the code and simplifies the logic ? I can then rebase your branch on top of it, if you'd like.

File: changelog/dpath.dd
Comment (WebFreak001) [reply changelog/dpath.dd:41]: this overrides the built-in dub location, so it would never even attempt to read anything from the built-in path, while for customCachePaths it's just an addition of paths.

File: source/dub/dub.d
Comment (WebFreak001) [reply source/dub/dub.d:329]: it's a little weird in the code, but I think it's great for user flexibility.

A use case here: package maintainer creates global settings.json in `/var/lib/dub` which then specifies where the user-overridable config is located. On linux distros that have read-only filesystems with some special paths for writable things that might for example be useful. Or for users using non-standard paths like storing all dub things on a separate flash drive or something that just isn't mounted in /etc.

Additionally it's possible to make the config only per-user and use an environment variable like `$HOME` inside the settings file so each user can have different dub settings.

File: changelog/dpath.dd
Comment (WebFreak001) [reply changelog/dpath.dd:23]: this overrides the user settings (e.g. what /etc/dub is for, but with a custom path)

`/var/lib/dub` is still used for reading the first base settings.

**Comment by WebFreak001** [comment]
> I just realized we read /etc/ before /var/lib...

I don't think that's the case, the systemSettings path is the very first config that is read, everything after that overrides the configs.

File: changelog/dpath.dd
Comment (WebFreak001) [reply changelog/dpath.dd:6]: that is for localRepository, but I think the more used thing is the userRepository. I don't think localRepository is ever used unless the user explicitly wants to store things in the local repository.

For simplification in the config, the config overrides both the user and the local repository.

File: source/dub/dub.d
Comment (WebFreak001) [reply source/dub/dub.d:297]: I think that's difficult as the state variables overrideUserSettings and overrideLocalRepository are later used to determine if custom path logic should be applied.

File: changelog/dpath.dd
Comment (Geod24) [reply changelog/dpath.dd:6]: `dub fetch` (and related `fetch` commands) use `localRepository`:
https://github.com/dlang/dub/blob/17d8c9467b81cf2b2ac57c338f01f71c2c71b013/source/dub/dub.d#L768-L773

The logic behind it, as I very recently understood and documented, is that using APPDATA for settings is good (you want your settings to roam), but not for packages (because it might use a lot of space / lead to complications with stale binaries):
https://github.com/dlang/dub/blob/17d8c9467b81cf2b2ac57c338f01f71c2c71b013/source/dub/dub.d#L1748-L1750

Comment (Geod24) [reply changelog/dpath.dd:6]: IOW:
```diff
-- Windows: `%APPDATA%/dub/{packages/,settings.json}`
+- Windows: `%APPDATA%/dub/settings.json` 
+- Windows: `%LOCALAPPDATA%/dub/packages/` 
```

I don't know if that should affect the functionality too.

**Comment by Geod24** [comment]
> I don't think that's the case, the systemSettings path is the very first config that is read, everything after that overrides the configs.

My bad, I meant the other way around :facepalm: 
I think we should read `/etc/` first, but we currently don't.

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:317]: What happens if the environment variable can't get expanded ?

File: changelog/dpath.dd
Comment (WebFreak001) [reply changelog/dpath.dd:6]: oh I see now, I thought localSettings correspond to PlacementLocation.local and userSettings to PlacementLocation.user, but it turns out to be a bit different there, which is confusing. Will update the changelog then.

File: source/dub/dub.d
Comment (Geod24) [reply source/dub/dub.d:1985]: Can we make things a bit more hierarchical, now that we have the ability to (easily) do so ?
Like `core.author` in `git`, you know.

File: source/dub/dub.d
Comment (WebFreak001) [reply source/dub/dub.d:317]: nothing is inserted in that case, (variable replaced with empty string) probably worth putting into dub-docs

Comment (WebFreak001) [reply source/dub/dub.d:317]: nothing is inserted in that case, (variable replaced with empty string) probably worth putting into dub-docs

Comment (Geod24) [reply source/dub/dub.d:317]: I think it should error out.

**Comment by WebFreak001** [comment]
> > I don't think that's the case, the systemSettings path is the very first config that is read, everything after that overrides the configs.
> 
> My bad, I meant the other way around facepalm I think we should read `/etc/` first, but we currently don't.

I strongly feel that we should not read /etc first, but /var/lib, like currently already done. I think the idea behind that is:

1. consistency with other programs (e.g. systemd)
2. the `/var/lib/...` path is maintained by the distro package maintainers and should never be updated by the user, this implies:
    - any setting in /var/lib should not override the user-set settings, thus it must not be loaded after the user settings in /etc

File: source/dub/dub.d
Comment (WebFreak001) [reply source/dub/dub.d:317]: done

**Comment by Geod24** [comment]
One use-case that might be important is the default placement path.
Currently the the `localRepository` (user HOME on POSIX) is the default placement.
Some users might want to use `system`, and some might want to use `local` (e.g. in the package itself). Currently the always have to pass an argument to DUB.

If you prefer we can tackle it in a separate PR, but for me it makes much more sense to think of them together (I thought the scope of this PR was originally bigger).

**Comment by WebFreak001** [comment]
I think a new setting to set the default placement location should be part of a separate PR.

**Comment by dk949** [comment]
I though you were planning to tie the location of `localRrpository` (aka `$HOME/.dub`, aka default placement) to `$DPATH`. If this is the case, wouldn't it make it somewhat inseparable from this PR?

(I too have possibly misunderstood the scope of these changes).

**Comment by WebFreak001** [comment]
rebased to new yaml loader

**Comment by WebFreak001** [comment]
> I though you were planning to tie the location of `localRrpository` (aka `$HOME/.dub`, aka default placement) to `$DPATH`. If this is the case, wouldn't it make it somewhat inseparable from this PR?
> 
> (I too have possibly misunderstood the scope of these changes).

no the default placement location setting I was talking about would set a default for the `dub --cache=*` setting (can set it to local, system or user)

**Comment by dk949** [comment]
Ohh, that makes sense. My bad.

**Comment by WebFreak001** [comment]
@Geod24 what's your opinion on this?

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:289]:
```suggestion
			string dubHome = environment.get("DUB_HOME");
			if (!dubHome.length) {
				if (auto dpath = environment.get("DPATH"))
					dubHome = (NativePath(dpath) ~ "dub/").toNativeString();
			}
```

IMO that makes more sense, no?

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:415]: Why the `override` ? Also wouldn't we be better simply exposing `SpecialDir` ?
But again, why expose it ? I don't see any user, and YAGNI.

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:311]:
```suggestion
		// Don't use it if either $DPATH or $DUB_HOME are set, as environment
		// variables usually take precedence over configuration.
```

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:307]:
```suggestion
		// Override user + local package path from system / binary settings
```

Since we already loaded the config from 3 locations

File: source/dub/dub.d
Comment (Geod24) [reply source/dub/dub.d:415]: Oh damn this is returning the config, not even the actual values! A big no from me.

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:322]: I'm a bit lost... Why do we have `dubHome` AND `userSettings` ?
Also use `this.m_config.userSettings.set` for condition, instead of checking length.

File: test/dpath-variable/dub.json
Comment (Geod24) [inline test/dpath-variable/dub.json:6]: Newline

File: test/dpath-variable/source/app.d
Comment (Geod24) [inline test/dpath-variable/source/app.d:3]: Newline

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:343]: Ah, that's why we have `dubHome`. And `userSettings`. And `localRepository`.
Honestly, just expose `dubHome`. YAGN `userSettings` / `localRepository` and it's not consistent with the env approach.

**Review by Geod24 — COMMENTED** [review]
The env approach is sound, the settings code still makes me uneasy but makes much more sense now, made a few suggestions.

File: source/dub/dub.d
Comment (WebFreak001) [reply source/dub/dub.d]: we can't do the `if (auto ...` thing here because we need to check for length, otherwise someone running `env DPATH= dub` would wrongly get dub home set to `NativePath("") ~ "dub/"`

**Comment by WebFreak001** [comment]
adjusted to review, can rebase if it's fine now

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:1883]: So those can be removed ?

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:225]: This only needs to be one boolean now, right ?

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:260]: No need to set it, right ?

File: source/dub/dub.d
Comment (Geod24) [inline source/dub/dub.d:275]: Ditto superfluous assign

File: changelog/dpath.dd
Comment (Geod24) [inline changelog/dpath.dd:28]: Changelog needs an update

**Comment by WebFreak001** [comment]
fixed left-overs

**Review by Geod24 — APPROVED** [review]

**Comment by WebFreak001** [comment]
thanks!

**Pull request merged**
